### ISSUE TITLE
Add optional error handling function to `Handler` trait

### ISSUE THREAD

**Issue by psFried** [psFried/roperator#5]
Operators may perform additional validation beyond what's possible with the structural validation in the CRD. Additionally, they may call out to external services. Both of those things significantly increase the chances of there being some sort of error when syncing a parent. 

We want to make it as easy as possible for operators to provide visibility to errors. Operators are typically used by people who don't (and shouldn't have to) understand all of the details of how everything works behind the scenes. It's a good practice to set a special `status` on the parent when an error occurs, because it allows users to easily see when an error occurs. Roperator currently supports this, but it does nothing to help make it easier.

Another good way to provide visibility to errors is emitting `Event`s. Unlike parent statuses, it's easy to have a generic process that watches only for events that indicate an error has occurred. Events also form a historical record of errors, which can be especially useful when there's transient issues. The parent status typically only shows the _current_ status, but it might not show a transient error that occurred in the past.

## Proposal

The idea is to add support for events and additionally to also allow operators to opt in to some reasonable automatic defaults for handling errors. This can be done by adding an additional function to the `Handler` trait that looks something like the following:

```rust
fn on_error(&self, description: &ErrorDescription) -> ErrorResponse;
```

The `ErrorDescription` and `ErrorResponse` types would be something like:

```rust
pub struct ErrorDescription<'a> {
    /// True if the error was returned from a `finalize` function, false if it was from `sync`
    pub is_finalize_error: bool,
    /// The sync request that was being handled when the error occurred
    pub parent: &'a roperator::handler::SyncRequest,
    pub error: roperator::error::Error, // a Box<dyn RoperatorError>, which supports downcasting
    /// The number of _consecutive_ errors
    pub error_count: u64,
}

pub struct ErrorResponse {
    /// how long to delay before retrying this particular parent
    pub retry_in: std::time::Duration,
    /// If event is Some, then roperator will emit an event that includes these details
    pub event: Option<EventDetails>,
    /// see below for an explanation of how this fields would be used
    pub status_error: Option<(String, serde_json::Value)>,
}

pub struct EventDetails {
    /// A short, machine-readable string describing the reason for the error event
    pub reason: String,
    /// A human-readable description of what happened
    pub message: String,
}
```

### Non-fallible function

The `on_error` function must not fail, or else we could easily end up in "turtles all the way down" scenario. There's just no reasonable way to try to handle errors in the error handler. The function signature does not allow for returning a `Result` in order to make this clear, and to discourage authors from making fallible function calls, e.g. HTTP calls to an external service.

### Error causes

Errors may be cause either by the `Handler` itself returning a `Result::Err` variant, or by an error that roperator encounters when communicating with the api server. Depending on the cause of an error, it may or may not be possible to emit events or set a parent status. For example, if there error is caused by a network error that prevents communication with the api server, or an RBAC misconfiguration that prevents roperator from creating events. We'll need to make it clear that this error handling is done on a best-effort basis. We should also support testing this error handling behavior in the `TestKit`.

### Parent Status

Given that errors may originate from outside the `Handler` itself (or even after the parent status has already been updated), it makes the semantics of the `status_error` field potentially confusing. The following sequence is one example of a potentially tricky situation:

1. The parent and all children are just fine and everything has been synced and is in a good state
2. Someone updates the `spec` of the parent, which triggers a sync.
3. The handler returns an error

The reason this is potentially ambiguous is that the status of the parent should continue to reflect the actual state of the child resources, and those resources may still be running just fine. But we want the status to also reflect the fact that we were not able to sync the latest changes to the `spec`. 

These type of scenarios imply that we should treat the status returned from the error handling function as _additive_. In other words, it should be added to the existing status instead of replacing it entirely. Thus, the `status_error` should not represent the entire status, but rather only the portion that describes the error. If `status_error` is Some, then it will contain a tuple that represents both the key and the value to add to the status. For example, returning `status_error: Some(("error".to_string(), Value::String("my error message".to_string()) )),` would cause the json `"error": "my error message"` to be added to the existing status instead of replacing it.

### Default Behavior

We want to make it as easy as possible to write operators that follow best practices, but in this case the events behavior should probably be opted into explicitly by the operator authors, since emitting events requires special permissions that may not be obvious. 

For the status, though, it may make sense to add a basic error message automatically. We already add the `observedGeneration` field to the status, so there's some precedent for that. And the risk or running into authorization errors is minimal, since we already update parent statuses as a matter of course.

**Comment by asthasr** [comment]
Some errors may not need a "retry," they should instead be marked as "fatal." Then, the parent object could be omitted from syncs until the fatal error is fixed.

**Comment by psFried** [comment]
Closing this, as it is mostly addressed by #25 et al. The one piece that isn't covered is emitting events, but I think that may be better off as a separate issue anyway.
